File: src/.fpm_spine_install.py
# Foundation Package Manager
# spine_install
# Developed by Brett. (https://github.com/notronaldmcdonald)

# imports

import subprocess
import tarfile
import pathlib
from pathlib import Path
import os
import sys
import string
from string import Template
from sys import platform

# some variables

user = os.getlogin()
core = "https://raw.githubusercontent.com/notronaldmcdonald/fpm/core/pkgs"
home = str(Path.home())

# USER - define your package sources by adding an additional dictionary entry. you can simply add additional 'curl' lines to the download portion of the script.
# Replace the selection at the end of the new curl line with your source name. (i.e. '[...] % core' > '[...] % userrepo')

# begin script

print("fpm: starting python script...")
exec(open('.target.txt').read())
print("fpm_spine: reading")
print("fpm_spine: operating system check")
if platform != "win32":
    print("fpm_spine: unix")
    print("fpm_spine: try to get package")
    try:
        # USER - this is where downloading happens. copy-paste the curl line and replace 'core' with the repo you added to the dictionary.
        confirmation = input("Continue? [Y/n] ")
        if confirmation == "n" or confirmation == "N":
            print("fpm_spine: aborted by user")
            sys.exit()
        else:
            # the aforementioned curl line is below
            subprocess.run(["curl", f"{core}/%s.tar.gz", "-o", f"{home}/.fpm/install/%s.tar.gz" % target])
            # the aforementioned curl line is above
            print("fpm_spine: unpacking")
            subprocess.run(["tar", "-xvzf", f"{home}/.fpm/install/%s.tar.gz" % target])
            exec(open("{home}/.fpm/install/fpkgd").read())
            print("fpm_spine: determine the binary")
            subprocess.run(["cp", f"{home}/.fpm/install/%s", f"{home}/.fpm/pkgs/%s" % base])
            print("fpm_spine: binary should be moved")
    except:
        print("fpm_spine: unknown exception occurred.")
    finally:
        print("fpm_spine: done that")
    print("fpm_spine: cleaning up now")
    subprocess.run(["rm", "-f", f"{home}/.fpm/install/*"])
    subprocess.run(["rm", "-f", ".target.txt"])
    print("fpm_spine: finished cleanup")
else:
    print("fpm_spine: windows")
    print("fpm_spine: try to get package")
    try:
        confirmation = input("Continue? [Y/n] ")
        if confirmation == "n" or confirmation == "N":
            print("fpm_spine: aborted by user")
            sys.exit()
        else:
            # same as the *nix section
            subprocess.run(["curl", f"{core}/%s.tar.gz", "-o", f"{home}\\.fpm\\install\\%s.tar.gz" % target])
            print("fpm_spine: unpacking")
            subprocess.run(["tar", "-xvzf", f"{home}\\.fpm\\install\\%s.tar.gz" % target])
            exec(open(f"{home}\\.fpm\\install\\fpkgd").read())
            print("fpm_spine: determine the binary")
            subprocess.run(["copy", f"{home}\\.fpm\\install\\%s", f"{home}\\.fpm\\pkgs\\%s" % base])
            print("fpm_spine: binary should be moved")
    except:
        print("fpm_spine: unknown exception occurred.")
    finally:
        print("fpm_spine: done that")
    print("fpm_spine: cleaning up now")
    subprocess.run(["del", "/f", f"{home}\\.fpm\\install\\*"])
    subprocess.run(["del", "/f", ".target.txt"])
print("fpm_spine: exit python")

File: src/.fpm_spine_install.rb
# Foundation Package Manager
# spine_install
# Developed by Brett. (https://github.com/notronaldmcdonald)

# dependencies

require 'fileutils'
require 'colorize'

# variables

home = Dir.home
op_sys = RUBY_PLATFORM

# repository targets
# USER - if you wish to add a repository, use this format.
core = "https://raw.githubusercontent.com/notronaldmcdonald/fpm/core/pkgs"
live = "https://raw.githubusercontent.com/notronaldmcdonald/fpm/live/pkgs"
mode = "r"
# target = "nnn" - this was a developer testing piece and it can be ignored.

# begin script

targetfile = File.open("/tmp/.target", mode)
target = targetfile.read
puts "Welcome!".green
puts "You requested: #{target}."
puts "Select a repository to download from:".yellow
puts "1) core (#{core})"
puts "2) live (#{live})"
# user can add more lines to this list
puts "Choose repository by entering the number.".yellow
target_repo = gets
target_repo = target_repo.chomp
puts "You chose: #{target_repo}"
# user can add more elsif blocks here
if target_repo == "1"
  puts "Trying to get #{target} from core.".blue
  system("curl #{core}/#{target}.tar.gz -o #{home}/.fpm/install/#{target}.tar.gz")
  puts "Verifying package intergrity..."
  system("sha256sum #{home}/.fpm/install/#{target}.tar.gz > /tmp/fpm_checksum")
  system("curl -fs #{core}/index/#{target}.hash -o /tmp/#{target}.hash")
  solid = FileUtils.identical?("/tmp/fpm_checksum", "/tmp/#{target}.hash")
  if solid == "false"
    puts "Unable to verify package integrity."
    puts "Install anyways? [y/N]"
    disregardcheck = gets
    disregardcheck = disregardcheck.chomp
    if disregardcheck == "y"
      puts "Disregard integrity check and try unpack".red
    else
      puts "Cancelling..."
      puts "Cleanup directories..."
      system("rm -f #{home}/.fpm/install/*")
      system("rm -f /tmp/fpm_* /tmp/*.hash")
      abort "fpm: fatal: package integrity could not be verified."
    end
  else
    puts "Package solid".green
  end
  puts "try unpack".blue
  system("tar -xzvf #{home}/.fpm/install/#{target}.tar.gz -C #{home}/.fpm/install")
  puts "If there was no errors, the package should have been unpacked.".yellow
  puts "try transfer".blue
  sleep(1)
  system("cp #{home}/.fpm/install/#{target} #{home}/.fpm/pkgs/#{target}")
  puts "If all went well, your package should be installed. If you haven't already, add #{home}/.fpm/pkgs to your PATH variable.".green
elsif target_repo == "2"
  puts "Note: packages in the 'live' repository are not subject to integrity checks.".red
  puts "This is because the code is often updated frequently and isn't intended to be mainstream."
  puts "Trying to get #{target} from live.".yellow
  system("curl #{live}/#{target}.tar.gz -o #{home}/.fpm/install/#{target}.tar.gz")
  puts "try unpack".blue
  system("tar -xzvf #{home}/.fpm/install/#{target}.tar.gz #{home}/.fpm/install")
  puts "If there was no errors, the package should have been unpacked.".yellow
  puts "try transfer".blue
  eval(File.read("fpkgd"), binding)
  system("cp #{home}/.fpm/install/#{target} #{home}/.fpm/pkgs/#{target}")
  puts "If all went well, your package should be installed. If you haven't already, add #{home}/.fpm/pkgs to your PATH variable.".green
else
  puts "unknown repo. exiting."
  abort "nonexistent repo.".red
end
targetfile.close
puts "fpm_spine: cleanup"
puts "Cleaning install directory..."
system("rm -f #{home}/.fpm/install/*")
puts "adding an entry to pkglist..."
system("echo #{target} >> #{home}/.fpm/pkgs/pkglist.txt")
puts "done.".green
